### PR TITLE
[web-browser] Rewrite Android code to Kotlin

### PR DESCRIPTION
**Pull request by barthap**
# Why

Resolves ENG-3061

# How

Auto convert + some manual tweaks. Took special care of null safety, it is not so obvious in this module. There are still areas to improve, but I'll take care of them when converting to expo-module.

# Test Plan

- NCL
- Native unit tests

### CODE REVIEW

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  @get:Throws(CurrentActivityNotFoundException::class)
+  private val currentActivity: Activity
+    get() {
+      val activityProvider = moduleRegistry.getModule(ActivityProvider::class.java)
+      return activityProvider?.currentActivity ?: throw CurrentActivityNotFoundException()
+    }
+
+  private fun createDefaultCustomTabsIntent(): Intent {
+    val builder = CustomTabsIntent.Builder()
+    val customTabsIntent = builder.build()
```
Comment (lukmccall) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt:79]:
```suggestion
    val customTabsIntent = CustomTabsIntent.Builder().build()
```
I would merge that

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    val intent = customTabsIntent.intent
+    intent.data = Uri.parse(DUMMY_URL)
+    return intent
```
Comment (lukmccall) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt:82]:
```suggestion
    return customTabsIntent.intent.apply {
      data = Uri.parse(DUMMY_URL)
    }
```

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    action = CustomTabsService.ACTION_CUSTOM_TABS_CONNECTION
+  }
+}
+
+private inline fun <T, R> Collection<T>.mapToDistinctArrayList(mapper: (T) -> R): ArrayList<R> {
+  val resultSet = LinkedHashSet<R>()
+  for (element in this) {
+    resultSet.add(mapper.invoke(element))
+  }
+  return ArrayList(resultSet)
```
Comment (lukmccall) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt:95]:
```suggestion
  return ArrayList(this.map { mapper(it) })
```

**Review by lukmccall — APPROVED** [review]

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  @ExpoMethod
+  fun warmUpAsync(packageName: String?, promise: Promise) {
+    try {
+      val resolvedPackageName = givenOrPreferredPackageName(packageName)
+      connectionHelper.warmUp(resolvedPackageName)
+      val result = Bundle().apply {
+        putString(SERVICE_PACKAGE_KEY, resolvedPackageName)
+      }
```
Comment (bbarthec) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt:56]: Take advantage of `bundleOf` coming from https://developer.android.com/kotlin/ktx/extensions-list

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  @ExpoMethod
+  fun coolDownAsync(packageName: String?, promise: Promise) {
+    try {
+      val resolvedPackageName = givenOrPreferredPackageName(packageName)
+      if (connectionHelper.coolDown(resolvedPackageName)) {
+        val result = Bundle().apply {
+          putString(SERVICE_PACKAGE_KEY, resolvedPackageName)
+        }
```
Comment (bbarthec) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt:70]:
```suggestion
        val result = bundleOf(
          SERVICE_PACKAGE_KEY to resolvedPackageName
        )
```

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt (hunk 4 above)
Comment (bbarthec) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt:56]: Take advantage of `bundleOf` from https://developer.android.com/kotlin/ktx/extensions-list#top-level_functions_5
```suggestion
      val result = bundleOf(
        SERVICE_PACKAGE_KEY to resolvedPackageName
      )
```

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+  @ExpoMethod
+  fun mayInitWithUrlAsync(url: String?, packageName: String?, promise: Promise) {
+    try {
+      val resolvedPackageName = givenOrPreferredPackageName(packageName)
+      // TODO (barthap): This might throw NPE when url is null
+      connectionHelper.mayInitWithUrl(resolvedPackageName, Uri.parse(url))
+      val result = Bundle().apply {
+        putString(SERVICE_PACKAGE_KEY, resolvedPackageName)
+      }
```
Comment (bbarthec) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt:88]:
```suggestion
      val result = bundleOf(
        SERVICE_PACKAGE_KEY to resolvedPackageName
      )
```

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+      // It might happen, that default activity does not support Chrome Tabs. Then it will be ResolvingActivity and we don't want to return it as a result.
+      val defaultCustomTabsPackage: String? = defaultPackage.takeIf { activities.contains(it) }
+
+      val result = Bundle().apply {
+        putStringArrayList(BROWSER_PACKAGES_KEY, activities)
+        putStringArrayList(SERVICE_PACKAGES_KEY, services)
+        putString(PREFERRED_BROWSER_PACKAGE, preferredPackage)
+        putString(DEFAULT_BROWSER_PACKAGE, defaultCustomTabsPackage)
+      }
```
Comment (bbarthec) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt:111]:
```suggestion
      val result = bundleOf(
        BROWSER_PACKAGES_KEY to activities,
        SERVICE_PACKAGES_KEY to services,
        PREFERRED_BROWSER_PACKAGE to preferredPackage,
        DEFAULT_BROWSER_PACKAGE to defaultCustomTabsPackage,
      )
```

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+  fun openBrowserAsync(url: String?, arguments: ReadableArguments, promise: Promise) {
+    val intent = createCustomTabsIntent(arguments)
+    intent.data = Uri.parse(url)
+    try {
+      if (customTabsResolver.canResolveIntent(intent)) {
+        customTabsResolver.startCustomTabs(intent)
+        val result = Bundle().apply {
+          putString("type", "opened")
+        }
+        promise.resolve(result)
```
Comment (bbarthec) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt:141]:
```suggestion
        val result = bundleOf(
          "type" to "opened"
        )
        promise.resolve(result)
```

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/CustomTabsActivitiesHelper.kt [hunk 9]
```diff
@@ -0,0 +1,27 @@
+package expo.modules.webbrowser
+
+import expo.modules.core.interfaces.InternalModule
+import expo.modules.webbrowser.error.PackageManagerNotFoundException
+import expo.modules.core.errors.CurrentActivityNotFoundException
+import android.content.Intent
+import java.util.ArrayList
+
+interface CustomTabsActivitiesHelper : InternalModule {
+  @get:Throws(PackageManagerNotFoundException::class, CurrentActivityNotFoundException::class)
```
Comment (bbarthec) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/CustomTabsActivitiesHelper.kt:10]: Do we have a policy about annotating exceptions types? 🤔 
I'm not sure what is the standard for Kotlin, but it seems as a good approach, let's add it to our codestyle guide, shall we?

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+
+private const val DUMMY_URL = "https://expo.dev"
+
+internal class InternalCustomTabsActivitiesHelper : CustomTabsActivitiesHelper {
+  private lateinit var moduleRegistry: ModuleRegistry
+
+  override fun onCreate(moduleRegistry: ModuleRegistry) {
+    this.moduleRegistry = moduleRegistry
+  }
+  override fun onDestroy() = Unit
```
Comment (bbarthec) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt:26]: Do you need it at all? 🤔

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt (hunk 3 above)
Comment (bbarthec) [reply packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt:95]: @lukmccall, isn't your suggestion dropping distinctness of elements from the resulting ArrayList? 🤔

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsConnectionHelper.kt [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+    }
+  }
+  // endregion
+
+  private fun ensureSession() {
+    if (!sessionActions.hasClient()) {
+      clientActions.executeOrQueueAction { client: CustomTabsClient ->
+        sessionActions.setClient(client.newSession(null))
+      }
+    }
```
Comment (bbarthec) [inline packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsConnectionHelper.kt:80]: Early returns for the win 🥇 
```suggestion
    if (sessionActions.hasClient()) {
      return
    }
    
    clientActions.executeOrQueueAction { client: CustomTabsClient ->
      sessionActions.setClient(client.newSession(null))
    }
```

**Review by bbarthec — APPROVED** [review]
Looking good and solidly migrated 💪 
My remarks:
- let's decide about annotating exception/error types in kotlin files - I see it might be useful to have them annotated. Let's add some rule to our codestyle guide.
- let's take advantage of Android utilities, especially from Jetpack/AndroidX ecosystem 👀 (`bundleOf`, etc)

Thank you for advocating step-by-step migration to `expo-modules` 🎉 🥳

**Comment by barthap** [comment]
- Regarding `bundleOf`: I'm definitely going to do it, but just wanted to postpone it to an expo-module migration PR, because some bundles might no longer be needed when taking advantage of built-in expo-module converters 😁 
- Regarding `@Throws` annotations: Actually they're only needed when calling Kotlin code from Java - the `@Throws` marks function as throwing Java checked exceptions, so it needs to be surrounded with `try-catch` in Java. As Kotlin doesn't have checked exceptions (for a reason btw), these annotations have a documentation-only role. I needed them only when migrating file-by-file to keep kotlinization safe. But, as I like to document, what kind of exceptions functions can throw, I don't mind replacing them with Javadoc comments, like I did e.g. here: 
 https://github.com/expo/expo/blob/74fe0b9159368b5b100840c5472898362b31a762/packages/expo-clipboard/android/src/main/java/expo/modules/clipboard/ClipboardImage.kt#L125-L130

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt (hunk 3 above)
Comment (barthap) [reply packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt:95]: It is 😉

**Comment by bbarthec** [comment]
> Regarding bundleOf: I'm definitely going to do it, but just wanted to postpone it to an expo-module migration PR, because some bundles might no longer be needed when taking advantage of built-in expo-module converters 😁

Ok, I expected you would respond like this 👍 

> Regarding @Throws annotations: Actually they're only needed when calling Kotlin code from Java - the @Throws marks function as throwing Java checked exceptions, so it needs to be surrounded with try-catch in Java. As Kotlin doesn't have checked exceptions (for a reason btw), these annotations have a documentation-only role. I needed them only when migrating file-by-file to keep kotlinization safe. But, as I like to document, what kind of exceptions functions can throw, I don't mind replacing them with Javadoc comments, like I did e.g. here:

I really this idea of adding them as comments 👍
Let's add this rule to the codestyle guide 🤩

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/DeferredClientActionsQueue.kt [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      action.apply(client)
+    } else {
+      addActionToQueue(action)
+    }
+  }
+
+  fun setClient(client: T) {
+    this.client = client
+    executeQueuedActions()
+  }
```
[the remark this answers is not in the document]
Comment (barthap) [reply packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/DeferredClientActionsQueue.kt:21]: Unfortunately, having `public set; private get` is impossible in Kotlin, and they don't plan this feature in the nearest future 😞 
https://youtrack.jetbrains.com/issue/KT-3110

File: packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt (hunk 10 above)
Comment (barthap) [reply packages/expo-web-browser/android/src/main/java/expo/modules/webbrowser/InternalCustomTabsActivitiesHelper.kt:26]: Yes, the `CustomTabsActivitiesHelper` implements unimodules `InternalModule` which requires them all. But I'll get rid of them anyway when migrating to expo-module

**Pull request merged**